src/mlpack/methods/reinforcement_learning/replay/hindsight_replay.hpp

```diff
@@ -0,0 +1,423 @@
+      herRatio(1),
+      goalSelectionStrategy(goalStrategy::FUTURE),
+      nSteps(0)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Construct an instance of hindsight experience replay class.
+   *
+   * @param batchSize Number of examples returned at each sample.
+   * @param capacity Total memory size in terms of number of examples.
+   * @param herRatio ratio of HER  to data coming from normal 
+   * experience replay in replay buffer
+   * @param nSteps Number of steps to look in the future.
+   * @param dimension The dimension of an encoded state.
+   * @param strategy goal selection startegy for HER
+   */
+  HindsightReplay(const size_t batchSize,
+                    const size_t capacity,
+                    const size_t herRatio = 4,
+                    EnvironmentType environment = EnvironmentType(),
+                    goalStrategy strategy = goalStrategy::FUTURE,
+                    const size_t nSteps = 1,
+                    const size_t dimension = StateType::dimension) :
+      batchSize(batchSize),
+      capacity((herRatio+1)*capacity),
+      position(0),
+      full(false),
+      herRatio(herRatio),
+      goalSelectionStrategy(strategy),
+      nSteps(nSteps),
+      states(dimension, (herRatio+1)*capacity),
+      actions((herRatio+1)*capacity),
+      rewards((herRatio+1)*capacity),
+      nextStates(dimension, (herRatio+1)*capacity),
+      goals(dimension, (herRatio+1)*capacity),
+      isTerminal((herRatio+1)*capacity),
+      environment(environment)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Sample goal according to goal strategy
+   * 
+   * @param transitionIndex index of current transition
+   */
+  void SampleGoal(StateType& desiredGoal, int transitionIndex)
+  {
+    if (goalSelectionStrategy == goalStrategy::FINAL)
+    {
+      // Sample goal as the final state of the epsiode
+      desiredGoal = episodeTransitions[episodeTransitions.size() -1].nextState;
+    }
+    else if (goalSelectionStrategy == goalStrategy::FUTURE)
+    {
+      // Sample goal as random state from that transition to end of epsiode
+      size_t index = arma::randi(
+        arma::distr_param(transitionIndex+1, episodeTransitions.size() -1));
+      desiredGoal = episodeTransitions[index].nextState;
+    }
+    else if (goalSelectionStrategy == goalStrategy::EPISODE)
+    {
+      // Sample goal as random state from the whole epsiode
+      desiredGoal = episodeTransitions[arma::randi(
+        arma::distr_param(0, episodeTransitions.size() -1))].nextState;
+    }
+  }
+
+  /**
+   * Sample goals according to goal strategy and HER ratio
+   * 
+   * @param desiredGoals desired goals as per goal strategy
+   * @param  transitionIndex index of current transition
+   */
+  void SampleGoals(std::vector<StateType>& desiredGoals, int transitionIndex)
+  { 
+    StateType desiredGoal;
+
+    // Store HER transitions based on strategy for each transition in epsiode
+    for (size_t goalIndex = 0; goalIndex < herRatio; ++goalIndex)
```

Comment:
When we say `ratio` its 0 => 1, so either the code is wrong or the description is.

Reply:
We could do this efficiently by first reserving spacing for `hindsight_transitions_per_regular_transition` aka `herRatio`.

```cpp
desiredGoals.reserve(desiredGoals.size() + herRatio);
// Then, we can push back using
std::generate_n(std::back_inserter(desiredGoals), herRatio,  SampleGoal);
```

Ofcourse its not that simple, but I guess you get the point?

Reply:
`SampleGoal` return type would be `StateType`, RVO optimisation should be able to deal with this.

Reply:
Actually ration means the number of HER Transition per normal per normal transitions. so if its 4. That means 4 for 1. Also ratio can be greater than 1. It just simply means how many times one number contains another.

Reply:
I know what it means, but not everyone will so its better to make the name more descriptive. Also look at other suggestions I've mentioned above

Reply:
> We could do this efficiently by first reserving spacing for `hindsight_transitions_per_regular_transition` aka `herRatio`.
> 
> ```c++
> std::vector<StateType> SampleGoals(const int transitionIdx)
> {
>    std::vector<StateType> desiredGoals(herRatio);
>    std::generate(desiredGoals.begin(), desiredGoals.end(), SampleGoal);
>    return desiredGoals;
> }
> ```
> 
> Ofcourse its not that simple, but I guess you get the point? This should be two line code.

I will take a look this.